### DISCUSSION TITLE
A couple of questions about pipeline communicate

### DISCUSSION THREAD

**Discussion by dvarrazzo** [psycopg/psycopg#435] [Q&A]
Hello @dlax!

I'm taking a look at the `pipeline_communicate()` generator. I have two questions:

1) The Python and the C implementation have a subtle difference, already introduced in 6ba95ce8f57462412e9c94162c555e3cbdb08cca: after receiving a sync, the Python version doesn't break, but keeps looping, I assume breaking only for the `is_busy()` check:

https://github.com/psycopg/psycopg/blob/5acb61464502dd7ca9ebd1c51b724ca472a4778d/psycopg/psycopg/generators.py#L210-L221

The C version, instead, breaks immediately:

https://github.com/psycopg/psycopg/blob/5acb61464502dd7ca9ebd1c51b724ca472a4778d/psycopg_c/psycopg_c/_psycopg/generators.pyx#L247-L262

They both seem to work ok, but is one version more correct than the other?

2) ISTM that this function is affected by something similar to #286. in particular, if we have already consumable results on the client (which can be checked using `is_busy()`), calling `consume_input()` (which hits the network) is superfluous. The fix for `fetch()` was in a05f4d9e3342843c5324d83987dbe16b22133a8e. The pipeline version seems to have a consume input (actually two - fixing one in af1b201a321b09ea6e3df5fe450b1f8926b72762) not guarded by `is_busy()`: maybe it can be improved?

Thank you very much!

**Comment by dlax** [comment]
> 1) The Python and the C implementation have a subtle difference, already introduced in 6ba95ce8f57462412e9c94162c555e3cbdb08cca: after receiving a sync, the Python version doesn't break, but keeps looping, I assume breaking only for the `is_busy()` check:
> 
> They both seem to work ok, but is one version more correct than the other?

I can't think of any reason to break specifically on receiving Sync(), so the Python version seems more correct.

> 2) ISTM that this function is affected by something similar to #286. in particular, if we have already consumable results on the client (which can be checked using `is_busy()`), calling `consume_input()` (which hits the network) is superfluous. The fix for `fetch()` was in a05f4d9e3342843c5324d83987dbe16b22133a8e. The pipeline version seems to have a consume input (actually two - fixing one in af1b201a321b09ea6e3df5fe450b1f8926b72762) not guarded by `is_busy()`: maybe it can be improved?

Not sure about this one, but maybe.

The implementation of `pipeline_communicate()` comes from [test_pipelined_insert in Postgres test suite](https://github.com/postgres/postgres/blob/34fa0ddae5cf27dd93a5855b30083185d4bb5a5b/src/test/modules/libpq_pipeline/libpq_pipeline.c#L743): this calls PQconsumeInput() at each iteration (select() call) and then loops while PQisBusy() to fetch results. This test is reproduced on our side as tests/scripts/pipeline-demo.py.

I just tried:
```diff
diff --git a/psycopg/psycopg/generators.py b/psycopg/psycopg/generators.py
index 584fe47b..cfa98f71 100644
--- a/psycopg/psycopg/generators.py
+++ b/psycopg/psycopg/generators.py
@@ -192,8 +192,9 @@ def _pipeline_communicate(
         ready = yield WAIT_RW
 
         if ready & READY_R:
-            pgconn.consume_input()
-            _consume_notifies(pgconn)
+            if pgconn.is_busy():
+                pgconn.consume_input()
+                _consume_notifies(pgconn)
 
             res: List[PGresult] = []
             while not pgconn.is_busy():
```
it seems to work (both test_pipeline.py and the demo script). Is that what you had in mind?

**Comment by dvarrazzo** [comment]
I am wondering if we can even guard the `Wait.RW` yielding on top of the function if we know that we already have data to process. If you look at `fetch()`, it jumps to `fetch_result()` without yielding at all. But in the pipeline code the things are made more complex by the write part.

I think a good test would be to look what happens with the single-row mode. Because the server delivers more than row to the client, and each one will be a separate result, if we `select()` between each row (as we were doing before fixing #286) then we are doing something wrong.

**Comment by dlax** [comment]
I don't think guarding the `Wait.RW` yielding would be very useful because the *main* point of `pipeline_communicate()` is to *send* commands, for which we need to `yield Wait.W`, and possibly/opportunistically read results if we also get `R-Ready`. In many cases, I think we'd not receive anything because we don't flush/sync at this point (which would happen when the user would call `fetch*()` eventually); this logic is mostly needed for cases where the output buffer of the server gets full (typically when inserting a lot as in the demo).

For the single-row mode, I suggested #409: this does not use `pipeline_communicate()` as commands sending and results fetch need to be done separately, and the latter uses the normal `fetch()` generator.
